contracts/golos.vesting/golos.vesting.cpp

```diff
@@ -0,0 +1,366 @@
+            item.return_date = time_point_sec(now() + OUTPUT_PAYOUT_PERIOD);
+//            item.return_date = time_point_sec(now()); // TODO test
+        });
+    }
+
+    tables::account_table account_recipient(_self, m_delegate_vesting.recipient);
+    auto balance_recipient = account_recipient.find(m_delegate_vesting.quantity.symbol.name());
+    if (balance_recipient != account_recipient.end()) {
+        account_recipient.modify(balance_recipient, m_delegate_vesting.sender, [&](auto &item){
+            item.delegate_vesting += m_delegate_vesting.quantity;
+        });
+    } else {
+        account_recipient.emplace(m_delegate_vesting.sender, [&](auto &item){
+            item.delegate_vesting = m_delegate_vesting.quantity;
+            item.vesting.symbol = m_delegate_vesting.quantity.symbol;
+            item.received_vesting.symbol = m_delegate_vesting.quantity.symbol;
+        });
+    }
+}
+
+void vesting::undelegate_vesting(const structures::transfer_vesting &m_transfer_vesting) {
+    tables::delegate_table table(_self, m_transfer_vesting.sender);
+    auto it_delegate = table.find(m_transfer_vesting.quantity.symbol.name());
+    eosio_assert(it_delegate != table.end(), "Not enough delegated vesting");
+    eosio_assert(it_delegate->return_date <= time_point_sec(now()), "Tokens are frozen until the end of the period");
+    eosio_assert(it_delegate->quantity >= m_transfer_vesting.quantity, "There are not enough delegated tools for output");
+
+    if (it_delegate->quantity == m_transfer_vesting.quantity) {
+        table.erase(it_delegate);
+    } else {
+        table.modify(it_delegate, m_transfer_vesting.sender, [&](auto &item){
+            item.quantity -= m_transfer_vesting.quantity;
+        });
+    }
+
+    _table_delegate_vesting.emplace(m_transfer_vesting.sender, [&](structures::return_delegate &item){
+        item.id = _table_delegate_vesting.available_primary_key();
+        item.recipient = m_transfer_vesting.sender;
+        item.amount = m_transfer_vesting.quantity;
+        item.date = time_point_sec(now() + OUTPUT_PAYOUT_PERIOD);
+    });
+
+    tables::account_table account_recipient(_self, m_transfer_vesting.recipient);
+    auto balance = account_recipient.find(m_transfer_vesting.quantity.symbol.name());
+    eosio_assert(balance != account_recipient.end(), "This token is not on the balance sheet");
+    account_recipient.modify(balance, m_transfer_vesting.sender, [&](auto &item){
+        item.delegate_vesting -= m_transfer_vesting.quantity;
+    });
+}
+
+void vesting::calculate_convert_vesting() {
+    require_auth(_self);
+
+    for (auto obj = _table_convert.begin(); obj != _table_convert.end(); ) {
+        if (obj->number_of_payments > 0 && (obj->payout_period > 0 && obj->payout_period <= OUTPUT_PAYOUT_PERIOD)) {
+            _table_convert.modify(obj, _self, [&](auto &item){
+                item.payout_period -= TIMEOUT;
```

Comment:
With each change of data, the `multi_index` code repacks them, which is very time consuming. Here, in fact, on every call the whole table will be repacked. It is much more efficient to store the time of the next withdrawal of vesting. Then you can build on it a secondary index and process only those requests, the time of which has come. Also, this implementation will be more reliable if there are problems with a deferred transaction.